extension/goext/types.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (C) 2017 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package goext
+
+import "encoding/json"
+
+type MaybeState int
+
+const (
+	MaybeUndefined MaybeState = iota
+	MaybeNull
+	MaybeNotNull
+)
+
+// Maybe is a 3-value state keeper
+//
+// Add it to your custom structure as a anonymous field to make structure 3-value
+// 3-value structures are marshalled into JSON value based on their state
+// Structures in:
+// - Undefined state won't appear in JSON at all
+// - Null state will appear in JSON as null
+// - NotNull state will appear in JSON as they would normally be marshalled
+//
+// Example:
+// Given structures:
+// type TestSuite struct {
+// 	 goext.Maybe
+//	 ID string `json:"id"`
+// }
+//
+// type Test struct {
+// 	 ID string `json:"id"`
+//	 Name string  `json:"name"`
+//	 TestSuite TestSuite `json:"test_suite"`
+// }
+// func MakeNullTestSuite() Test {
+// 	 return TestSuite{
+//   Maybe: goext.Maybe{MaybeState: goext.MaybeNull}
+//	 }
+// }
+//
+// json.Marshal(env.Util().ResourceToMap(&Test{ID: "some-id", Name:"My resource"}))
+// Will produce {"id": "some-id", "name": "My resource"}
+//
+// json.Marshal(env.Util().ResourceToMap(&Test{ID: "some-id", Name:"My resource"}, TestSuite: MakeNullTestSuite()))
+// Will produce {"id": "some-id", "name": "My resource", "test_suite": null}
+//
+// json.Marshal(env.Util().ResourceToMap(&Test{ID: "some-id", Name:"My resource"}, TestSuite: MakeTestSuite("test-suite-id")))
+// Will produce {"id": "some-id", "name": "My resource", "test_suite": {"id": "test-suite-id"}}
+//
+//
+// Structures embedding Maybe are sensitive to internal state changes.
+// User must take extra care to change MaybeState field during every operation on any field.
+// Example:
+// nullTestSuite := MakeNullTestSuite()
+// nullTestSuite.ID = "some-test-id"
+//
+// nullTestSuite.MaybeState is still MaybeNull, it will be marshalled into null even that it contain proper value
+//
+
+type Maybe struct {
+	MaybeState MaybeState
+}
+
+// IsUndefined returns whether value is undefined
+func (m Maybe) IsUndefined() bool {
+	return m.MaybeState == MaybeUndefined
+}
+
+// IsNull returns whether value is null
+func (m Maybe) IsNull() bool {
+	return m.MaybeState == MaybeNull
+}
+
+// IsNotNull returns whether value is defined and not null
+func (m Maybe) IsNotNull() bool {
+	return m.MaybeState == MaybeNotNull
```

Comment:
In the case of undefined, this return false, right? Is it really convenient to strictly distinguish null and undefined in Go side?

Reply:
Yes, it will return false.  It is convenient, since JSON has 3 state values. Developer should take care of it, because API of other application may has different interpretation of undefined and null. If case doesn't apply, developer have choice to not care about difference.

Reply:
Ok, then I think a kind of `IsValid` feature for this MaybeX type. I do not want to write so many `!val.IsNll() && !val.IsUndefined()`.